config.org
#+TITLE: GNU Emacs Config
#+AUTHOR: Joaquim Gregório
#+DESCRIPTION: My personal Emacs config.

* Important! Put this in your init.el 
For the first time you must save *config.org* file to be generated a *config.el* file. Since the auto-tangle is enabled, any change you do in the *config.org* file will rewrite the *config.el* file.
#+begin_example emacs-lisp
(defconst dot-emacs (concat (getenv "HOME") "/.emacs.d/" "config.el")
  "My dot emacs file")

(require 'bytecomp)
(setq compiled-dot-emacs (byte-compile-dest-file dot-emacs))

(if (or (not (file-exists-p compiled-dot-emacs))
	(file-newer-than-file-p dot-emacs compiled-dot-emacs)
        (equal (nth 4 (file-attributes dot-emacs)) (list 0 0)))
    (load dot-emacs)
  (load compiled-dot-emacs))

(add-hook 'kill-emacs-hook
          '(lambda () (and (file-newer-than-file-p dot-emacs compiled-dot-emacs)
                           (byte-compile-file dot-emacs))))

(defun loadup-gen ()
  "Generate the lines to include in the lisp/loadup.el file
to place all of the libraries that are loaded by your InitFile
into the main dumped emacs"
  (interactive)
  (defun get-loads-from-*Messages* ()
    (save-excursion
      (let ((retval ()))
	(set-buffer "*Messages*")
	(beginning-of-buffer)
	(while (search-forward-regexp "^Loading " nil t)
	  (let ((start (point)))
	    (search-forward "...")
	    (backward-char 3)
	    (setq retval (cons (buffer-substring-no-properties start (point)) retval))))
	retval))))

(loadup-gen)

;; time the loading of the .emacs
;; keep this on top of your .emacs
(defvar *emacs-load-start* (current-time))
(defun anarcat/time-to-ms (time)
  (+ (* (+ (* (car time) (expt 2 16)) (car (cdr time))) 1000000) (car (cdr (cdr time)))))
(defun anarcat/display-timing ()
  (message ".emacs loaded in %fms" (/ (- (anarcat/time-to-ms (current-time)) (anarcat/time-to-ms *emacs-load-start*)) 1000000.0)))
(add-hook 'after-init-hook 'anarcat/display-timing t)
#+end_example
* Load First
** Startup performance
#+begin_src emacs-lisp :tangle yes
;; Minimize garbage collection during startup
(setq gc-cons-threshold most-positive-fixnum) 

;; Increase the amount of data which Emacs reads from the process#
;; Again the emacs default is too low 4k considering that the some of the language server responses are in 800k - 3M range.
(setq read-process-output-max (* 1024 1024)) ;; 1mb

;; Lower threshold back to 8 MiB (default is 800kB)
(add-hook 'emacs-startup-hook
          (lambda ()
            (setq gc-cons-threshold (expt 2 23))))

(defun efs/display-startup-time ()
(message "Emacs loaded in %s with %d garbage collections."
        (format "%.2f seconds"
                (float-time
                    (time-subtract after-init-time before-init-time)))
        gcs-done))

(add-hook 'emacs-startup-hook #'efs/display-startup-time)
#+end_src
** Initialize Package Sources 
#+begin_src emacs-lisp :tangle yes
;; Initialize package sources
(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                        ("org" . "https://orgmode.org/elpa/")
                        ("elpa" . "https://elpa.gnu.org/packages/")))
(package-initialize)
(unless package-archive-contents
(package-refresh-contents))

;; Initialize use-package on non-linux platforms
;;(unless (package-installed-p 'use-package)
;;(package-install 'use-package))

(require 'use-package)
(setq use-package-always-ensure t)
#+end_src

** Files
#+begin_src emacs-lisp :tangle yes
;; make backup to a designated dir, mirroring the full path
(defun my-backup-file-name (fpath)
  "Return a new file path of a given file path.
If the new path's directories does not exist, create them."
  (let* (
        (backupRootDir "~/.emacs.d/emacs-backup/")
        (filePath (replace-regexp-in-string "[A-Za-z]:" "" fpath )) ; remove Windows driver letter in path, ➢ for example: “C:”
        (backupFilePath (replace-regexp-in-string "//" "/" (concat backupRootDir filePath "~") ))
        )
    (make-directory (file-name-directory backupFilePath) (file-name-directory backupFilePath))
    backupFilePath
  )
)

(setq make-backup-file-name-function 'my-backup-file-name)
;; stop creating #autosave# files
(setq auto-save-default nil)
#+end_src
** Evil Mode

#+begin_src emacs-lisp :tangle yes
  (use-package evil
  :defer t
  :init      ;; tweak evil's configuration before loading it
  (setq evil-want-integration t) ;; This is optional since it's already set to t by default.
  (setq evil-want-keybinding nil)
  (setq evil-vsplit-window-right t)
  (setq evil-split-window-below t)
  (evil-mode))
  (use-package evil-tutor)
  (use-package evil-collection
    :defer t
    :after (evil)
    :config
    (setq evil-collection-mode-list '(dired)) ;; for dired work properly
    (evil-collection-init))
#+end_src

* UI
** Dashboard
#+begin_src emacs-lisp :tangle yes
;; If the dashboard doesn`t word after init, create ~/Org/{Tasks.org,Birthdays.org,Habits.org} files
(use-package dashboard
  :defer nil
  :ensure t
  :config
  (setq dashboard-items '((recents . 5)
                    (projects . 4)
                    (bookmarks . 3))
      dashboard-startup-banner 'logo
      dashboard-center-content t
      dashboard-set-navigator t
      dashboard-show-shortcuts nil
      dashboard-set-heading-icons t
      dashboard-set-file-icons t
      dashboard-projects-switch-function 'counsel-projectile-switch-project-by-name
      dashboard-week-agenda nil)
  (dashboard-setup-startup-hook))
  ;; Mandatory to get dashboard when using `emacsclient -c`
  (setq initial-buffer-choice (lambda () (get-buffer "*dashboard*")))
#+end_src
** Centaur tabs
#+begin_src emacs-lisp :tangle yes
  (use-package centaur-tabs
    ;;:defer t
    :ensure t
    :config
       (setq centaur-tabs-style "bar"
              centaur-tabs-height 24
              centaur-tabs-set-icons t
              centaur-tabs-set-modified-marker t
              centaur-tabs-close-button "✕"
              centaur-tabs-modified-marker "•"
              centaur-tabs-set-bar 'over
              x-underline-at-descent-line t)
       (centaur-tabs-headline-match)
       ;; (setq centaur-tabs-gray-out-icons 'buffer)
       ;; (centaur-tabs-enable-buffer-reordering)
       ;; (setq centaur-tabs-adjust-buffer-order t)
       (centaur-tabs-mode t)
       (setq uniquify-separator "/")
       (setq uniquify-buffer-name-style 'forward)
       (defun centaur-tabs-buffer-groups ()
         "`centaur-tabs-buffer-groups' control buffers' group rules.

     Group centaur-tabs with mode if buffer is derived from `eshell-mode' `emacs-lisp-mode' `dired-mode' `org-mode' `magit-mode'.
     All buffer name start with * will group to \"Emacs\".
     Other buffer group by `centaur-tabs-get-group-name' with project name."
         (list
          (cond
            ;; ((not (eq (file-remote-p (buffer-file-name)) nil))
            ;; "Remote")
            ((or (string-equal "*" (substring (buffer-name) 0 1))
                 (memq major-mode '(magit-process-mode
                                    magit-status-mode
                                    magit-diff-mode
                                    magit-log-mode
                                    magit-file-mode
                                    magit-blob-mode
                                    magit-blame-mode
                                    )))
             "Emacs")
            ((derived-mode-p 'prog-mode)
             "Editing")
            ((derived-mode-p 'dired-mode)
             "Dired")
            ((memq major-mode '(helpful-mode
                                help-mode))
             "Help")
            ((memq major-mode '(org-mode
                                org-agenda-clockreport-mode
                                org-src-mode
                                org-agenda-mode
                                org-beamer-mode
                                org-indent-mode
                                org-bullets-mode
                                org-cdlatex-mode
                                org-agenda-log-mode
                                diary-mode))
             "OrgMode")
            (t
             (centaur-tabs-get-group-name (current-buffer))))))
    :hook
    (dashboard-mode . centaur-tabs-local-mode)
    (term-mode . centaur-tabs-local-mode)
    (vterm-mode . centaur-tabs-local-mode)
    (calendar-mode . centaur-tabs-local-mode)
    (org-mode . centaur-tabs-local-mode)
    (org-agenda-mode . centaur-tabs-local-mode)
    (helpful-mode . centaur-tabs-local-mode)
    :bind
    ("C-<prior>" . centaur-tabs-backward)
    ("C-<next>" . centaur-tabs-forward)
    ("C-c t s" . centaur-tabs-counsel-switch-group)
    ("C-c t p" . centaur-tabs-group-by-projectile-project)
    ("C-c t g" . centaur-tabs-group-buffer-groups)
    (:map evil-normal-state-map
           ("g t" . centaur-tabs-forward)
           ("g T" . centaur-tabs-backward)))

    (global-set-key (kbd "C-x t e") #'centaur-tabs-mode)
#+end_src
** hl-todo
#+begin_src emacs-lisp :tangle yes
    (use-package hl-todo
      :defer t
      :config
      (setq hl-todo-highlight-punctuation ":"
            hl-todo-keyword-faces
            `(;; For things that need to be done, just not today.
              ("TODO" warning bold)
              ;; For problems that will become bigger problems later if not
              ;; fixed ASAP.
              ("FIXME" error bold)
              ;; For tidbits that are unconventional and not intended uses of the
              ;; constituent parts, and may break in a future update.
              ("HACK" font-lock-constant-face bold)
              ;; For things that were done hastily and/or hasn't been thoroughly
              ;; tested. It may not even be necessary!
              ("REVIEW" font-lock-keyword-face bold)
              ;; For especially important gotchas with a given implementation,
              ;; directed at another user other than the author.
              ("NOTE" success bold)
              ;; For things that just gotta go and will soon be gone.
              ("DEPRECATED" font-lock-doc-face bold)
              ;; For a known bug that needs a workaround
              ("BUG" error bold)
              ;; For warning about a problematic or misguiding code
              ("XXX" font-lock-constant-face bold))))

  (global-hl-todo-mode)
#+end_src
** Font Configuration

#+begin_src emacs-lisp :tangle yes
  (defun efs/set-font-faces ()
    (message "Setting faces!")
    (set-face-attribute 'default nil :font "JetBrainsMono NF" :weight 'regular :height 105)
    (set-face-attribute 'bold nil :font "JetBrainsMono NF" :weight 'bold)
    (set-face-attribute 'italic nil :font "JetBrainsMono NF" :weight 'light :slant 'italic)
 
    ;; Set the fixed pitch face
    (set-face-attribute 'fixed-pitch nil :font "JetBrainsMono NF" :height 105)

    ;; Set the variable pitch face
    (set-face-attribute 'variable-pitch nil :font "JetBrainsMono NF" :height 105))

  (if (daemonp)
          (add-hook 'after-make-frame-functions
                (lambda (frame)
                (setq doom-modeline-icon t)
                (with-selected-frame frame
                    (efs/set-font-faces))))
      (efs/set-font-faces))
#+end_src

** Ui Improvements

#+begin_src emacs-lisp :tangle yes
  ;; (setq inhibit-startup-message t) ; don't show startup screen
  (scroll-bar-mode -1) ; Disable visible scrollbar
  (tool-bar-mode -1) ; Disable the toolbar
  (tooltip-mode -1) ; Disable tooltips
  (set-fringe-mode 10) ; Give some breathing room
  (menu-bar-mode -1) ; Disable the menu bar
  (setq visible-bell t) ; Set up the visible bell
  (setq scroll-conservatively 101)

  ;; Disable all GUI popups and tooltips
  (setq use-dialog-box nil)

  ;; Wrap words
  (add-hook 'text-mode-hook 'visual-line-mode)

  ;; Set frame transparency
  ;;(set-frame-parameter (selected-frame) 'alpha '(95 . 95))
  ;;(add-to-list 'default-frame-alist `(alpha . ,'(95 . 95)))
  (set-frame-parameter (selected-frame) 'fullscreen 'maximized)
  (add-to-list 'default-frame-alist '(fullscreen . maximized))

  ;; Line numbers
  (column-number-mode)
  (global-display-line-numbers-mode t)
  ;; Disable line numbers for some modes
  (dolist (mode '(org-mode-hook
                  term-mode-hook
                  vterm-mode-hook
                  shell-mode-hook
                  treemacs-mode-hook
                  eshell-mode-hook))
    (add-hook mode (lambda () (display-line-numbers-mode 0))))

  ;; Cursor shape
  (setq-default cursor-type 'hbar)
  (set-cursor-color "#ffffff") ;; or box, hollow, hbar

  ;; DocView
  ;; (setq doc-view-continuous t)
  #+end_src

** Theme
#+begin_src emacs-lisp :tangle yes
;;Set up some theme
(use-package doom-themes
  ;;:defer t
  :ensure t
  :config
  ;; Global settings (defaults)
  (setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
	  doom-themes-enable-italic t) ; if nil, italics is universally disabled
  (load-theme 'doom-one t)

  ;; Enable flashing mode-line on errors
  (doom-themes-visual-bell-config)
  ;; Enable custom neotree theme (all-the-icons must be installed!)
  (doom-themes-neotree-config)
  ;; or for treemacs users
  (setq doom-themes-treemacs-theme "doom-atom") ; use "doom-colors" for less minimal icon theme
  (doom-themes-treemacs-config)
  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))
#+end_src
*** Doom
#+begin_src emacs-lisp :tangle yes
  (use-package all-the-icons :defer t :ensure t)

  ;;Doom mode line
  (use-package doom-modeline
    :defer t
    :ensure t
    :init (doom-modeline-mode 1))
#+end_src

** Yafolding
#+begin_src emacs-lisp :tangle yes
  (use-package yafolding :defer t)
  (defvar yafolding-mode-map
  (let ((map (make-sparse-keymap)))
    (define-key map (kbd "<C-S-return>") #'yafolding-hide-parent-element)
    (define-key map (kbd "<C-M-return>") #'yafolding-toggle-all)
    (define-key map (kbd "<C-return>") #'yafolding-toggle-element)
    map))
#+end_src
** Emoji
#+begin_src emacs-lisp :tangle yes
  (use-package emojify
    :defer t
    :hook (after-init . global-emojify-mode))
#+end_src
* Org mode
Org Mode is one of the hallmark features of Emacs. It is a rich document editor, project planner, task and time tracker, blogging engine, and literate coding utility all wrapped up in one package.
** Better Font Faces
The efs/org-font-setup function configures various text faces to tweak the sizes of headings and use variable width fonts in most cases so that it looks more like we’re editing a document in org-mode. We switch back to fixed width (monospace) fonts for code blocks and tables so that they display correctly.
 #+begin_src emacs-lisp :tangle yes
   (defun efs/org-font-setup ()
   ;; Replace list hyphen with dot
   (font-lock-add-keywords 'org-mode
			   '(("^ *\\([-]\\) "
			      (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "•"))))))

   ;; Set faces for heading levels
   (dolist (face '((org-level-1 . 1.2)
		   (org-level-2 . 1.1)
		   (org-level-3 . 1.05)
		   (org-level-4 . 1.0)
		   (org-level-5 . 1.1)
		   (org-level-6 . 1.1)
		   (org-level-7 . 1.1)
		   (org-level-8 . 1.1)))
     (set-face-attribute (car face) nil :font "FiraCode NF" :height (cdr face)))

   ;; Ensure that anything that should be fixed-pitch in Org files appears that way
   (set-face-attribute 'org-block nil    :foreground nil :inherit 'fixed-pitch)
   (set-face-attribute 'org-table nil    :inherit 'fixed-pitch)
   (set-face-attribute 'org-formula nil  :inherit 'fixed-pitch)
   (set-face-attribute 'org-code nil     :inherit '(shadow fixed-pitch))
   (set-face-attribute 'org-table nil    :inherit '(shadow fixed-pitch))
   (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
   (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
   (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
   (set-face-attribute 'org-checkbox nil  :inherit 'fixed-pitch)
   (set-face-attribute 'line-number nil :inherit 'fixed-pitch)
   (set-face-attribute 'line-number-current-line nil :inherit 'fixed-pitch))
  (defun efs/org-mode-setup ()
  (org-indent-mode)
  (variable-pitch-mode 1)
  (visual-line-mode 1))
#+end_src
** Basic Config
This section contains the basic configuration for org-mode plus the configuration for Org agendas and capture templates.
#+begin_src emacs-lisp :tangle yes
  (use-package org
    ;;:defer t
    :pin org
    :commands (org-capture org-agenda)
    :hook (org-mode . efs/org-mode-setup)
    :config
    (setq org-ellipsis " ▾")

    (setq org-agenda-start-with-log-mode t)
    (setq org-log-done 'time)
    (setq org-log-into-drawer t)

    (setq org-agenda-files
          '("~/Org/Tasks.org"
            "~/Org/Habits.org"
            "~/Org/Birthdays.org"))

    (require 'org-habit)
    (add-to-list 'org-modules 'org-habit)
    (setq org-habit-graph-column 60)

    (setq org-todo-keywords
      '((sequence "TODO(t)" "NEXT(n)" "|" "DONE(d!)")
        (sequence "BACKLOG(b)" "PLAN(p)" "READY(r)" "ACTIVE(a)" "REVIEW(v)" "WAIT(w@/!)" "HOLD(h)" "|" "COMPLETED(c)" "CANC(k@)")))

    (setq org-refile-targets
      '(("Archive.org" :maxlevel . 1)
        ("Tasks.org" :maxlevel . 1)))

    ;; Save Org buffers after refiling!
    (advice-add 'org-refile :after 'org-save-all-org-buffers)

    (setq org-tag-alist
      '((:startgroup)
         ; Put mutually exclusive tags here
         (:endgroup)
         ("@errand" . ?E)
         ("@home" . ?H)
         ("@work" . ?W)
         ("agenda" . ?a)
         ("planning" . ?p)
         ("publish" . ?P)
         ("batch" . ?b)
         ("note" . ?n)
         ("idea" . ?i)))

    ;; Configure custom agenda views
    (setq org-agenda-custom-commands
     '(("d" "Dashboard"
       ((agenda "" ((org-deadline-warning-days 7)))
        (todo "NEXT"
          ((org-agenda-overriding-header "Next Tasks")))
        (tags-todo "agenda/ACTIVE" ((org-agenda-overriding-header "Active Projects")))))

      ("n" "Next Tasks"
       ((todo "NEXT"
          ((org-agenda-overriding-header "Next Tasks")))))

      ("W" "Work Tasks" tags-todo "+work-email")

      ;; Low-effort next actions
      ("e" tags-todo "+TODO=\"NEXT\"+Effort<15&+Effort>0"
       ((org-agenda-overriding-header "Low Effort Tasks")
        (org-agenda-max-todos 20)
        (org-agenda-files org-agenda-files)))

      ("w" "Workflow Status"
       ((todo "WAIT"
              ((org-agenda-overriding-header "Waiting on External")
               (org-agenda-files org-agenda-files)))
        (todo "REVIEW"
              ((org-agenda-overriding-header "In Review")
               (org-agenda-files org-agenda-files)))
        (todo "PLAN"
              ((org-agenda-overriding-header "In Planning")
               (org-agenda-todo-list-sublevels nil)
               (org-agenda-files org-agenda-files)))
        (todo "BACKLOG"
              ((org-agenda-overriding-header "Project Backlog")
               (org-agenda-todo-list-sublevels nil)
               (org-agenda-files org-agenda-files)))
        (todo "READY"
              ((org-agenda-overriding-header "Ready for Work")
               (org-agenda-files org-agenda-files)))
        (todo "ACTIVE"
              ((org-agenda-overriding-header "Active Projects")
               (org-agenda-files org-agenda-files)))
        (todo "COMPLETED"
              ((org-agenda-overriding-header "Completed Projects")
               (org-agenda-files org-agenda-files)))
        (todo "CANC"
              ((org-agenda-overriding-header "Cancelled Projects")
               (org-agenda-files org-agenda-files)))))))

    (setq org-capture-templates
      `(("t" "Tasks / Projects")
        ("tt" "Task" entry (file+olp "~/Org/Tasks.org" "Inbox")
             "* TODO %?\n  %U\n  %a\n  %i" :empty-lines 1)

        ("j" "Journal Entries")
        ("jj" "Journal" entry
             (file+olp+datetree "~/Org/Journal.org")
             "\n* %<%I:%M %p> - Journal :journal:\n\n%?\n\n"
             ;; ,(dw/read-file-as-string "~/Notes/Templates/Daily.org")
             :clock-in :clock-resume
             :empty-lines 1)
        ("jm" "Meeting" entry
             (file+olp+datetree "~/Org/Journal.org")
             "* %<%I:%M %p> - %a :meetings:\n\n%?\n\n"
             :clock-in :clock-resume
             :empty-lines 1)

        ("w" "Workflows")
        ("we" "Checking Email" entry (file+olp+datetree "~/Org/Journal.org")
             "* Checking Email :email:\n\n%?" :clock-in :clock-resume :empty-lines 1)

        ("m" "Metrics Capture")
        ("mw" "Weight" table-line (file+headline "~/Org/Metrics.org" "Weight")
         "| %U | %^{Weight} | %^{Notes} |" :kill-buffer t)))

    (define-key global-map (kbd "C-c j")
      (lambda () (interactive) (org-capture nil "jj")))

    (efs/org-font-setup))
 #+end_src
** Nicer Heading Bullets
org-bullets replaces the heading stars in org-mode buffers with nicer looking characters that you can control.
#+begin_src emacs-lisp :tangle yes
  (use-package org-bullets
    ;;:defer t
    :hook (org-mode . org-bullets-mode))
    ;; :custom
    ;; (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))
#+end_src
** Center Org Buffers

We use visual-fill-column to center org-mode buffers for a more pleasing writing experience as it centers the contents of the buffer horizontally to seem more like you are editing a document. This is really a matter of personal preference so you can remove the block below if you don’t like the behavior.
#+begin_src emacs-lisp :tangle yes
  (defun efs/org-mode-visual-fill ()
    (setq visual-fill-column-width 100
          visual-fill-column-center-text t)
    (visual-fill-column-mode 1))

  (use-package visual-fill-column
    ;;:defer t
    :ensure t
    :hook (org-mode . efs/org-mode-visual-fill))
#+end_src
** Configure Babel Languages
To execute or export code in org-mode code blocks, you’ll need to set up org-babel-load-languages for each language you’d like to use. This page documents all of the languages that you can use with org-babel.
#+begin_src emacs-lisp :tangle yes
(with-eval-after-load 'org
  (org-babel-do-load-languages
      'org-babel-load-languages
      '((emacs-lisp . t)
      (python . t)
      (C . t)))

  (push '("conf-unix" . conf-unix) org-src-lang-modes))
#+end_src
** Structure Templates
Org Mode’s structure templates feature enables you to quickly insert code blocks into your Org files in combination with org-tempo by typing < followed by the template name like el or py and then press TAB. For example, to insert an empty emacs-lisp block below, you can type <el and press TAB to expand into such a block.

You can add more src block templates below by copying one of the lines and changing the two strings at the end, the first to be the template name and the second to contain the name of the language as it is known by Org Babel.
#+begin_src emacs-lisp :tangle yes
(with-eval-after-load 'org
  ;; This is needed as of Org 9.2
  (require 'org-tempo)

  (add-to-list 'org-structure-template-alist '("sh" . "src shell"))
  (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
  (add-to-list 'org-structure-template-alist '("py" . "src python"))
  (add-to-list 'org-structure-template-alist '("cc" . "src C")))
#+end_src
** Auto-tangle Configuration Files
This snippet adds a hook to org-mode buffers so that efs/org-babel-tangle-config gets executed each time such a buffer gets saved. This function checks to see if the file being saved is the Emacs.org file you’re looking at right now, and if so, automatically exports the configuration here to the associated output files.
#+begin_src emacs-lisp :tangle yes
;; Automatically tangle our Emacs.org config file when we save it
(defun efs/org-babel-tangle-config ()
  (when (string-equal (file-name-directory (buffer-file-name))
                      (expand-file-name user-emacs-directory))
    ;; Dynamic scoping to the rescue
    (let ((org-confirm-babel-evaluate nil))
      (org-babel-tangle))))

(add-hook 'org-mode-hook (lambda () (add-hook 'after-save-hook #'efs/org-babel-tangle-config)))
#+end_src
** Org Mime
#+begin_src emacs-lisp :tangle yes
(use-package org-mime :ensure t)
#+end_src
** Hide Source Blocks
#+begin_src emacs-lisp :tangle yes
;; Org mode
(add-hook 'org-mode-hook
        '(lambda ()
            ;; Hide org block lines

            ;; Unset any previous customization for the background color
            (set-face-attribute 'org-block-begin-line nil :background 'unspecified)
            (set-face-attribute 'org-block-end-line nil :background 'unspecified)

            ;; Set the foreground color to the value of the background color
            (set-face-attribute 'org-block-begin-line nil
            :foreground (face-background 'org-block-begin-line nil 'default))
            (set-face-attribute 'org-block-end-line nil
            :foreground (face-background 'org-block-end-line nil 'default))
            ))
#+end_src
* Keybindings

#+begin_src emacs-lisp :tangle yes
;; Make ESC quit prompts
(global-set-key (kbd "<escape>") 'keyboard-escape-quit)
;; Adjust text scale
;;(global-set-key (kbd "C--") 'text-scale-decrease)
;;(global-set-key (kbd "C-=") 'text-scale-increase)

;; See keybindings
(use-package command-log-mode)
#+end_src

* Ivy

#+begin_src emacs-lisp :tangle yes
  ;; Use Ivy and Counsel for completions
  (use-package ivy
    ;;:defer t
    :diminish
    :bind (("C-s" . swiper)
           :map ivy-minibuffer-map
           ("TAB" . ivy-alt-done)	
           ("C-l" . ivy-alt-done)
           ("C-j" . ivy-next-line)
           ("C-k" . ivy-previous-line)
           :map ivy-switch-buffer-map
           ("C-k" . ivy-previous-line)
           ("C-l" . ivy-done)
           ("C-d" . ivy-switch-buffer-kill)
           :map ivy-reverse-i-search-map
           ("C-k" . ivy-previous-line)
           ("C-d" . ivy-reverse-i-search-kill))
    :config
    (ivy-mode 1)
    (setq ivy-use-virtual-buffers t)
    (setq ivy-wrap t)
    (setq ivy-count-format "(%d/%d) ")
    (setq enable-recursive-minibuffers t)

    ;; Use different regex strategies per completion command
    (push '(completion-at-point . ivy--regex-fuzzy) ivy-re-builders-alist)
    (push '(swiper . ivy--regex-ignore-order) ivy-re-builders-alist)
    (push '(counsel-M-x . ivy--regex-ignore-order) ivy-re-builders-alist)

    ;; Set minibuffer height for different commands
    (setf (alist-get 'swiper ivy-height-alist) 15)
    (setf (alist-get 'counsel-switch-buffer ivy-height-alist) 7))

  (use-package counsel
    ;;:defer t
    :init
    :bind (("M-x" . counsel-M-x)
           ;; ("C-x b" . counsel-ibuffer)
           ("C-x b" . counsel-switch-buffer)
           ("C-x C-f" . counsel-find-file)
           ;; ("C-M-j" . 'counsel-switch-buffer)
           ("C-M-j" . 'counsel-ibuffer)
           :map minibuffer-local-map
           ("C-r" . 'counsel-minibuffer-history))
    :custom
    (counsel-linux-app-format-function #'counsel-linux-app-format-function-name-only)
    :config
    (counsel-mode 1)
    (setq ivy-initial-inputs-alist nil))  ;; Don't start searches with ^

   ;; Ivy rich
   (use-package ivy-rich
     ;;:defer t
     :init
     (ivy-rich-mode 1))
#+end_src

* Which key
#+begin_src emacs-lisp :tangle yes
  ;; Which key
  (use-package which-key
    ;;:defer t
    :init (which-key-mode)
    :diminish which-key-mode
    :config
    (setq which-key-idle-delay 0.4))
#+end_src
* Usefull Plugins
** Helpful

#+begin_src emacs-lisp :tangle yes
  (use-package helpful
    ;;:defer t
    :custom
    (counsel-describe-function-function #'helpful-callable)
    (counsel-describe-variable-function #'helpful-variable)
    :bind
    ([remap describe-function] . counsel-describe-function)
    ([remap describe-command] . helpful-command)
    ([remap describe-variable] . counsel-describe-variable)
    ([remap describe-key] . helpful-key))
#+end_src

* Language Server Protocol (LSP)
  With above setup done, below we will setup several packages closely related to LSP.
** Run Code
We can use quickrun package to execute code (if it has main). E.g. If you have a java file with main method, it will run with the associated shortcut key ~C-c r~ or quickrun command. Quickrun has support for several languages.
#+begin_src emacs-lisp :tangle yes
  (use-package quickrun 
  ;;:defer t
  :bind ("C-c r" . quickrun))
#+end_src

** Yasnippet

config.el
;; Initialize package sources
(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
			 ("org" . "https://orgmode.org/elpa/")
			 ("elpa" . "https://elpa.gnu.org/packages/")))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))

;; Initialize use-package on non-linux platforms
(unless (package-installed-p 'use-package)
  (package-install 'use-package))

(require 'use-package)
(setq use-package-always-ensure t)

(use-package evil
  :init      ;; tweak evil's configuration before loading it
  (setq evil-want-integration t) ;; This is optional since it's already set to t by default.
  (setq evil-want-keybinding nil)
  (setq evil-vsplit-window-right t)
  (setq evil-split-window-below t)
  (evil-mode))
(use-package evil-tutor)

(setq inhibit-startup-message t) ; don't show startup screen

(scroll-bar-mode -1) ; Disable visible scrollbar
(tool-bar-mode -1) ; Disable the toolbar
(tooltip-mode -1) ; Disable tooltips
(set-fringe-mode 10) ; Give some breathing room
(menu-bar-mode -1) ; Disable the menu bar

;; Set up the visible bell
(setq visible-bell t)
;; Set up the font
(set-face-attribute 'default nil :font "JetBrainsMono Nerd Font" :height 130)
;; Set up some theme
(use-package doom-themes
  :ensure t
  :config
  ;; Global settings (defaults)
  (setq doom-themes-enable-bold t    ; if nil, bold is universally disabled
	doom-themes-enable-italic t) ; if nil, italics is universally disabled
  (load-theme 'doom-one t)

  ;; Enable flashing mode-line on errors
  (doom-themes-visual-bell-config)
  ;; Enable custom neotree theme (all-the-icons must be installed!)
  (doom-themes-neotree-config)
  ;; or for treemacs users
  (setq doom-themes-treemacs-theme "doom-atom") ; use "doom-colors" for less minimal icon theme
  (doom-themes-treemacs-config)
  ;; Corrects (and improves) org-mode's native fontification.
  (doom-themes-org-config))

;; Line numbers
(column-number-mode)
(global-display-line-numbers-mode t)
;; Disable line numbers for some modes
(dolist (mode '(org-mode-hook
		term-mode-hook
		shell-mode-hook
		eshell-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 0))))

(use-package neotree)

;; Make ESC quit prompts
(global-set-key (kbd "<escape>") 'keyboard-escape-quit)

;; See keybindings
(use-package command-log-mode)

;; Use Ivy and Counsel for completions
(use-package ivy
  :diminish
  :bind (("C-s" . swiper)
	 :map ivy-minibuffer-map
	 ("TAB" . ivy-alt-done)	
	 ("C-l" . ivy-alt-done)
	 ("C-j" . ivy-next-line)
	 ("C-k" . ivy-previous-line)
	 :map ivy-switch-buffer-map
	 ("C-k" . ivy-previous-line)
	 ("C-l" . ivy-done)
	 ("C-d" . ivy-switch-buffer-kill)
	 :map ivy-reverse-i-search-map
	 ("C-k" . ivy-previous-line)
	 ("C-d" . ivy-reverse-i-search-kill))
  :config
  (ivy-mode 1)
  (setq ivy-use-virtual-buffers t)
  (setq ivy-wrap t)
  (setq ivy-count-format "(%d/%d) ")
  (setq enable-recursive-minibuffers t)

  ;; Use different regex strategies per completion command
  (push '(completion-at-point . ivy--regex-fuzzy) ivy-re-builders-alist)
  (push '(swiper . ivy--regex-ignore-order) ivy-re-builders-alist)
  (push '(counsel-M-x . ivy--regex-ignore-order) ivy-re-builders-alist)

  ;; Set minibuffer height for different commands
  (setf (alist-get 'swiper ivy-height-alist) 15)
  (setf (alist-get 'counsel-switch-buffer ivy-height-alist) 7))

(use-package counsel
  :init
  :bind (("M-x" . counsel-M-x)
	 ("C-x b" . counsel-ibuffer)
	 ("C-x C-f" . counsel-find-file)
	 ("C-M-j" . 'counsel-switch-buffer)
	 :map minibuffer-local-map
	 ("C-r" . 'counsel-minibuffer-history))
  :custom
  (counsel-linux-app-format-function #'counsel-linux-app-format-function-name-only)
  :config
  (counsel-mode 1)
  (setq ivy-initial-inputs-alist nil))  ;; Don't start searches with ^

 ;; Ivy rich
 (use-package ivy-rich
   :init
   (ivy-rich-mode 1))

;; Doom mode line
(use-package doom-modeline
  :ensure t
  :init (doom-modeline-mode 1)
  :custom ((doom-modeline-height 12)))

;; Delimiters
(use-package rainbow-delimiters
  :hook (prog-mode . rainbow-delimiters-mode))

;; Which key
(use-package which-key
  :init (which-key-mode)
  :diminish which-key-mode
  :config
  (setq which-key-idle-delay 0.4))

(use-package helpful
  :ensure t
  :custom
  (counsel-describe-function-function #'helpful-callable)
  (counsel-describe-variable-function #'helpful-variable)
  :bind
  ([remap describe-function] . counsel-describe-function)
  ([remap describe-command] . helpful-command)
  ([remap describe-variable] . counsel-describe-variable)
  ([remap describe-key] . helpful-key))

(use-package lsp-mode)
